feat(scan): add expect method to wait for certain issue by severity threshold

ostridm opened the pull request:
closes #50

ostridm committed:
feat(scan): add expect method to wait for certain issue by severity threshold

closes #50

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +96,19 @@ public async Task Stop()
+    cancellationTokenSource.CancelAfter(_options.Timeout);
+
+    var predicate = (Scan _) =>
+      Task.FromResult(_scanState.IssuesBySeverity?.Any(x => x.Number > 0 && x.Type == expectation) ?? false);
```
Severity is a threshold describing a range from x to max, the strict equality won't' work as expected. For details see https://github.com/NeuraLegion/sectester-js/blob/7b409959bb74fbdf00a8179517d81c90f5fd3208/packages/scan/src/models/Severity.ts#L7

ostridm committed:
feat(scan): add expect method to wait for certain issue by severity threshold

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/90#discussion_r1042290938

closes #50

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +104,20 @@ public async Task Stop()
+
+    var predicate = () =>
+      _scanState.IssuesBySeverity?.Any(x => x.Number > 0 &&
+      SeverityRanges.Any(y => expectation == y.Key && y.Value.Contains(x.Type))) ?? false;
```
Why don't move this into a separate method? 

```csharp
public bool IsInExpectedSeverityRange(Severity severity) {
  if (_scanState.IssuesBySeverity == null) return false;
  return _scanState.IssuesBySeverity.Any(x => SeverityRanges.Any(y => expectation == y.Key && y.Value.Contains(x.Type)));
}
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -460,4 +460,59 @@ public async Task Expect_PredicateThrows_DoesNothing()
+
+  [Theory]
+  [MemberData(nameof(ActiveStatuses))]
+  public async Task Expect_GivenSeverityThresholdWithConditionSatisfied_Returns(ScanStatus scanStatus)
```
```suggestion
  public async Task Expect_ConditionSatisfied_Returns(ScanStatus scanStatus)
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -460,4 +460,59 @@ public async Task Expect_PredicateThrows_DoesNothing()
+  }
+  
+  [Fact]
+  public async Task Expect_GivenSeverityThresholdAndCancelledToken_Returns()
```
```suggestion
  public async Task Expect_CancellationTokenIsCancelled_Returns()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -460,4 +460,59 @@ public async Task Expect_PredicateThrows_DoesNothing()
   }
+  
+  [Fact]
+  public async Task Expect_GivenSeverityThresholdCancelledByTimeout_Returns()
```
```suggestion
  public async Task Expect_CancelledByTimeout_Returns()
```

ostridm committed:
feat(scan): add expect method to wait for certain issue by severity threshold

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/90#discussion_r1043506415
https://github.com/NeuraLegion/sectester-net/pull/90#discussion_r1043512465
https://github.com/NeuraLegion/sectester-net/pull/90#discussion_r1043513093
https://github.com/NeuraLegion/sectester-net/pull/90#discussion_r1043514021

closes #50

ostridm committed:
feat(scan): add expect method to wait for certain issue by severity threshold

fix flaky tests

closes #50

derevnjuk committed:
Merge branch 'master' into feat_#50/extend-scan-by-adding-a-expect-method-to-wait-for-certain-issue-by-severity-threshold

derevnjuk commented:
@ostridm is it ready for review?
